src/System.Memory/src/System/MemoryExtensions.cs

```diff
@@ -119,6 +119,24 @@ public static int SequenceCompareTo<T>(this Span<T> first, ReadOnlySpan<T> secon
             return SpanHelpers.SequenceCompareTo(ref MemoryMarshal.GetReference(first), first.Length, ref MemoryMarshal.GetReference(second), second.Length);
         }
 
+        /// <summary>
+        /// Reverses the sequence of the elements in the entire span.
+        /// </summary>
+        public static void Reverse<T>(this Span<T> span)
+        {
+            ref T p = ref MemoryMarshal.GetReference(span);
+            int i = 0;
+            int j = span.Length - 1;
+            while (i < j)
+            {
+                T temp = Unsafe.Add(ref p, i);
+                Unsafe.Add(ref p, i) = Unsafe.Add(ref p, j);
+                Unsafe.Add(ref p, j) = temp;
```

Comment:
Does this end up calculating each of the locations twice, or does the JIT see and remove the duplication?

Reply:
The JIT does remove one of them, but not both (see the lea instruction). Although, the calculations are done as part of the mov/lea instructions.

```text
00007ff8`72c60940 488b01          mov     rax,qword ptr [rcx] ds:00000008`d577e178=0000018ccee62c88
00007ff8`72c60943 33d2            xor     edx,edx
00007ff8`72c60945 8b4908          mov     ecx,dword ptr [rcx+8]
00007ff8`72c60948 ffc9            dec     ecx
00007ff8`72c6094a 85c9            test    ecx,ecx
00007ff8`72c6094c 7e21            jle     ConsoleApp16!Program.Reverse[[System.Int32, System.Private.CoreLib]](System.Span`1<Int32>)+0x2f (00007ff8`72c6096f)
00007ff8`72c6094e 4c63c2          movsxd  r8,edx
00007ff8`72c60951 468b0c80        mov     r9d,dword ptr [rax+r8*4]
00007ff8`72c60955 4e8d0480        lea     r8,[rax+r8*4]
00007ff8`72c60959 4c63d1          movsxd  r10,ecx
00007ff8`72c6095c 468b1c90        mov     r11d,dword ptr [rax+r10*4]
00007ff8`72c60960 458918          mov     dword ptr [r8],r11d
00007ff8`72c60963 46890c90        mov     dword ptr [rax+r10*4],r9d
00007ff8`72c60967 ffc2            inc     edx
00007ff8`72c60969 ffc9            dec     ecx
00007ff8`72c6096b 3bd1            cmp     edx,ecx
00007ff8`72c6096d 7cdf            jl      ConsoleApp16!Program.Reverse[[System.Int32, System.Private.CoreLib]](System.Span`1<Int32>)+0xe (00007ff8`72c6094e)
00007ff8`72c6096f c3              ret
```

**But, we could re-write it as follows to produce optimal disassembly (courtesy of Levi):**
```C#
// Version A
public static unsafe void Reverse<T>(Span<T> span)
{
    if (span.Length > 1)
    {
        ref T p = ref MemoryMarshal.GetReference(span);
        IntPtr i = IntPtr.Zero;
        IntPtr j = (IntPtr)span.Length - 1;

        do
        {
            var temp1 = Unsafe.Add(ref p, i);
            var temp2 = Unsafe.Add(ref p, j);
            Unsafe.Add(ref p, i) = temp2;
            Unsafe.Add(ref p, j) = temp1;

            i += 1;
            j -= 1;
        } while (i.ToPointer() < j.ToPointer());
    }
}
```
```text
00007fff`98502740 488b01          mov     rax,qword ptr [rcx] ds:0000003f`fa97dbb8=0000022fca27ae90
00007fff`98502743 8b5108          mov     edx,dword ptr [rcx+8]
00007fff`98502746 83fa01          cmp     edx,1
00007fff`98502749 7e25            jle     EndianReverse!EndianReverse.Program.Reverse[[System.Byte, System.Private.CoreLib]](System.Span`1<Byte>)+0x30 (00007fff`98502770)
00007fff`9850274b 33c9            xor     ecx,ecx
00007fff`9850274d 4863d2          movsxd  rdx,edx
00007fff`98502750 48ffca          dec     rdx
00007fff`98502753 440fb60408      movzx   r8d,byte ptr [rax+rcx]
00007fff`98502758 440fb60c10      movzx   r9d,byte ptr [rax+rdx]
00007fff`9850275d 44880c08        mov     byte ptr [rax+rcx],r9b
00007fff`98502761 44880410        mov     byte ptr [rax+rdx],r8b
00007fff`98502765 48ffc1          inc     rcx
00007fff`98502768 48ffca          dec     rdx
00007fff`9850276b 483bca          cmp     rcx,rdx
00007fff`9850276e 72e3            jb      EndianReverse!EndianReverse.Program.Reverse[[System.Byte, System.Private.CoreLib]](System.Span`1<Byte>)+0x13 (00007fff`98502753)
00007fff`98502770 c3              ret
```

**However, for some reason, the following (sub-optimal) implementation ends up doing slightly better in the micro-benchmarks (especially for Guids):**
```C#
// Version B
public static unsafe void Reverse<T>(this Span<T> span)
{
    if (span.Length > 1)
    {
        ref T p = ref MemoryMarshal.GetReference(span);
        IntPtr i = IntPtr.Zero;
        IntPtr j = (IntPtr)span.Length - 1;

        do
        {
            T temp = Unsafe.Add(ref p, i);
            Unsafe.Add(ref p, i) = Unsafe.Add(ref p, j);
            Unsafe.Add(ref p, j) = temp;

            i += 1;
            j -= 1;
        } while (i.ToPointer() < j.ToPointer());
    }
}
```

```text
00007fff`98512740 488b01          mov     rax,qword ptr [rcx] ds:000000d5`db37dd08=00000218b916ae90
00007fff`98512743 8b5108          mov     edx,dword ptr [rcx+8]
00007fff`98512746 83fa01          cmp     edx,1
00007fff`98512749 7e28            jle     EndianReverse!EndianReverse.Program.Reverse[[System.Byte, System.Private.CoreLib]](System.Span`1<Byte>)+0x33 (00007fff`98512773)
00007fff`9851274b 33c9            xor     ecx,ecx
00007fff`9851274d 4863d2          movsxd  rdx,edx
00007fff`98512750 48ffca          dec     rdx
00007fff`98512753 440fb60408      movzx   r8d,byte ptr [rax+rcx]
00007fff`98512758 4c8d0c08        lea     r9,[rax+rcx]
00007fff`9851275c 440fb61410      movzx   r10d,byte ptr [rax+rdx]
00007fff`98512761 458811          mov     byte ptr [r9],r10b
00007fff`98512764 44880410        mov     byte ptr [rax+rdx],r8b
00007fff`98512768 48ffc1          inc     rcx
00007fff`9851276b 48ffca          dec     rdx
00007fff`9851276e 483bca          cmp     rcx,rdx
00007fff`98512771 72e0            jb      EndianReverse!EndianReverse.Program.Reverse[[System.Byte, System.Private.CoreLib]](System.Span`1<Byte>)+0x13 (00007fff`98512753)
00007fff`98512773 c3              ret
```

![image](https://user-images.githubusercontent.com/6527137/35084230-d2479782-fbd7-11e7-9af2-a254876d49a3.png)

Should I change it to the version that produces the most optimal codegen? I can't really explain the micro-benchmark results that I am seeing though since they conflict with the disassembly.

cc @GrabYourPitchforks, @AndyAyersMS

Reply:
Presumably the codegen for Guids looks differently different; not just that it has an extra lea =>r9 over the "A" version

Reply:
I would stick with the original, especially if we've used that idiomatic sequence for swapping array elements elsewhere. Best not to have too many variant patterns for the same thing if we can help it. 

If you think the codegen for that version can be improved, please open an issue over in CoreCLR.

Reply:
Filed an issue and keeping the implementation unchanged:
https://github.com/dotnet/coreclr/issues/15937